doc/manual/src/contributing/hacking.md

```diff
@@ -99,8 +99,70 @@ You can run the whole testsuite with `make check`, or the tests for a specific c
 ### Functional tests
 
 The functional tests reside under the `tests` directory and are listed in `tests/local.mk`.
-The whole testsuite can be run with `make install && make installcheck`.
-Individual tests can be run with `make tests/{testName}.sh.test`.
+Each test is a bash script.
+
+The whole testsuite can be run with:
+```shell-session
+$ make install && make installcheck
+ran test tests/foo.sh... [PASS]
+ran test tests/bar.sh... [PASS]
+...
+```
+
+Individual tests can be run with:
+```shell-session
+$ make tests/${testName}.sh.test
+ran test tests/${testName}.sh... [PASS]
+```
+or without `make`, with:
+```shell-session
+$ ./mk/run_test.sh tests/${testName}.sh
+ran test tests/${testName}.sh... [PASS]
+```
+
+To see the output unsuppressed, one can also run
+```shell-session
+$ ./mk/debug_test.sh tests/${testName}.sh
++ foo
+output from foo
++ bar
+output from bar
+...
+```
+The test script will be traced with `set -x` and the output displayed as it happens, regardless of whether the test succeeds or fails.
+
+#### Debugging failing functional tests
+
+When a functional test fails, one is quite lucky if it does so on the first line.
+More commonly, the test failure somewhere in the middle of the script.
+
+To figure out what's wrong, it is convenient to run the test up to the failing nix command, and then run that test with a debugger like GDB.
+
+For example, if the script looks like:
+```bash
+foo
+nix blah blub
+bar
+```
+on would could edit it like so:
+```diff
+ foo
+-nix blah blub
++gdb --args nix blah blub
+ bar
+```
+
+Then, when one runs the script with `./mk/debug_test.sh`, it will drop them into GDB once the script reaches that:
+```shell-session
+$ ./mk/debug_test.sh tests/${testName}.sh
```

Comment:
Maybe rename to `debug-test.sh` (dash instead of underscore)? We don't really use underscores in the Nix project.

Reply:
Maybe (?) we could also provide a `Makefile` target for that to be consistent with running the test. So that command would be something like `make test.${testName}.sh.debug`

Reply:
Both done.